src/main/java/sx/blah/discord/util/MessageHistoryRange.java

```diff
@@ -0,0 +1,280 @@
+			for (int i = 0; i < cached.size(); i++) {
+				if (!checkStart(cached.get(i))) // if we just went past start message (confusing logic)
+					return i == 0 ? null : cached.get(i - 1);
+			}
+		} else {
+			for (int i = 0; i < cached.size(); i++) {
+				if (checkStart(cached.get(i))) // if we just hit start message (confusing logic)
+					return cached.get(i);
+			}
+		}
+
+		// If execution has reached this point, need to request history to find message
+		long last = cached.isEmpty() ? 0 : cached.get(cached.size() - 1).getLongID(); // start at last element of cached stuff
+
+		// Get chunk of messages
+		long fLast = last;
+		IMessage[] chunk = RequestBuffer.request(() -> {
+			return channel.requestHistory(fLast, null, Channel.MESSAGE_CHUNK_COUNT);
+		}).get();
+
+		// While we haven't reached the channel's end
+		while (chunk.length != 0) {
+
+			// Same logic as for cache
+			if (chronological) {
+				for (int i = 0; i < chunk.length; i++) {
+					if (!checkStart(chunk[i])) // if we just went past start message (confusing logic)
+						return i == 0 ? null : chunk[i - 1];
+				}
+			} else {
+				for (int i = 0; i < chunk.length; i++) {
+					if (checkStart(chunk[i])) // if we just hit start message (confusing logic)
+						return chunk[i];
+				}
+			}
+
+			last = chunk[chunk.length - 1].getLongID();
+			long lastF = last;
+			chunk = RequestBuffer.request(() -> {
+				return channel.requestHistory(lastF, null, Channel.MESSAGE_CHUNK_COUNT);
+			}).get();
+		}
+
+		return null;
+	}
+
+	/**
+	 * Checks whether a message is beyond the starting point.
+	 * NOTE: This method accounts for the direction of `end` relative to `start`.
+	 *
+	 * @param msg the message to check
+	 * @return whether the message is past the starting point.
+	 */
+	public boolean checkStart(IMessage msg) {
+		return chronological ?
+				start != Endpoint.NOW &&
+						msg.getTimestamp().compareTo(start.getTime()) > (start.inclusive ? -1 : 0) :
+				start != Endpoint.CHANNEL_CREATE &&
+						msg.getTimestamp().compareTo(start.getTime()) < (start.inclusive ? 1 : 0);
+	}
+
+	/**
+	 * Checks whether a message is before the ending point.
+	 * NOTE: This method accounts for the direction of `end` relative to `start`.
+	 *
+	 * @param msg the message to check
+	 * @return whether the message is past the ending point.
+	 */
+	public boolean checkEnd(IMessage msg) {
+		return chronological ?
+				end != Endpoint.CHANNEL_CREATE &&
+						msg.getTimestamp().compareTo(end.getTime()) < (end.inclusive ? 1 : 0) :
+				end != Endpoint.NOW &&
+						msg.getTimestamp().compareTo(end.getTime()) > (end.inclusive ? -1 : 0);
+	}
+
+	/**
+	 * This class represents an ending point for a {@link MessageHistoryRange}. It can be defined as either a timestamp
+	 * or a message, and includes whether or not the point should be included in the range. This class also includes
+	 * constants for the idea of a channel's beginning and a channel's most recent point.
+	 */
+	public static class Endpoint {
```

Comment:
I .. just realized the conflict there. I'll do a quick refactor to Point instead. Since the class has to be called as MessageHistory.Point, it should be clear what it represents, and I won't change any documentation to say "point" instead of "endpoint" to maintain clarity.